File: admin/src/shared/api/admin-prompt-safety.ts
// 作者: 杨永的Agent
// 日期: 2026-05-20
// 修改功能: Admin 安全防护聚合读 API（ai.prompt-safety · FE_HANDOFF）

import { HTTPError } from 'ky'

import type { PromptPackListResponse, PromptPackSummary } from '@/shared/api/admin-prompt-packs'
import { AppError } from '@/shared/api/errors'
import { httpClient, toAppError } from '@/shared/api/http-client'

const PREFIX = 'v1/admin/prompt-safety'

export type SafetyInterceptCategory = 'runtime' | 'prompt' | 'tool' | 'session'
export type RuntimeGovernanceHrefKind = 'confirmation' | 'policy' | 'routing'
export type ToolPolicyKind = 'deny' | 'allowlist' | 'shadow'
export type ToolEnforcementMode = 'default_deny' | 'allowlist_only' | 'approve_then_allow'
export type SessionPolicyStatus = 'enabled' | 'warning' | 'info'

export interface SafetyBlocklistRule {
  ruleId: string
  phrase: string
  matchMode: string
}

export interface SafetyBlocklistResponse {
  safetyPhraseBlocklistRevision: string
  safetyPhraseScanScopeDefault: string
  rules: SafetyBlocklistRule[]
}

export interface SafetyOverview {
  safetyPhraseBlocklistRevision: string
  safetyPhraseScanScopeDefault: string
  platformSafetyPack: PromptPackSummary | null
  safetyPromptPackCount: number
  publishedSafetyPackCount: number
  enabledConfirmationRulesCount: number
  globalAgentSwitchOn: boolean
  opsSuspended: boolean
}

export interface RuntimeSafetyGovernanceRow {
  key: string
  name: string
  riskLevelLabel: string
  autoExecSummary: string
  confirmSummary: string
  breakerSummary: string
  hrefKind: RuntimeGovernanceHrefKind
}

export interface RuntimeSafetyGovernanceResponse {
  items: RuntimeSafetyGovernanceRow[]
}

export interface ToolSafetyPolicyRow {
  key: string
  toolPattern: string
  policy: ToolPolicyKind
  enforcementMode: ToolEnforcementMode
  note: string
}

export interface ToolSafetyPolicyResponse {
  items: ToolSafetyPolicyRow[]
}

export interface SessionSafetyPolicyItem {
  key: string
  label: string
  summary: string
  status: SessionPolicyStatus
}

export interface SessionSafetyPolicyResponse {
  items: SessionSafetyPolicyItem[]
}

export interface SafetyInterceptItem {
  id: string
  ts: string
  category: SafetyInterceptCategory
  scenarioLabel: string
  kindLabel: string
  reason: string
  subject: string | null
  executionId: string | null
  matchedRuleId: string | null
}

export interface SafetyInterceptListResponse {
  items: SafetyInterceptItem[]
  total: number
}

export interface ListSafetyInterceptsQuery {
  category?: SafetyInterceptCategory
  limit?: number
  offset?: number
}

async function normalizeKyError(error: unknown): Promise<never> {
  if (error instanceof HTTPError) {
    const cloned = error.response.clone()
    const body: unknown = await cloned.json().catch(() => null)
    if (
      body !== null &&
      typeof body === 'object' &&
      'detail' in body &&
      typeof (body as { detail: unknown }).detail === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { detail: string }).detail, {
        status: error.response.status,
        code,
      })
    }
  }
  throw await toAppError(error)
}

export function explainPromptSafetyError(e: unknown, fallback = '请求失败'): string {
  if (!(e instanceof AppError)) return fallback
  return e.message || fallback
}

export async function getPromptSafetyOverview(): Promise<SafetyOverview> {
  try {
    return await httpClient.get(`${PREFIX}/overview`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getPromptSafetyBlocklist(): Promise<SafetyBlocklistResponse> {
  try {
    return await httpClient.get(`${PREFIX}/blocklist`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function listPromptSafetyPacks(): Promise<PromptPackListResponse> {
  try {
    return await httpClient.get(`${PREFIX}/prompt-packs`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getPromptSafetyRuntimeGovernance(): Promise<RuntimeSafetyGovernanceResponse> {
  try {
    return await httpClient.get(`${PREFIX}/runtime-governance`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getPromptSafetyToolPolicies(): Promise<ToolSafetyPolicyResponse> {
  try {
    return await httpClient.get(`${PREFIX}/tool-policies`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function getPromptSafetySessionPolicy(): Promise<SessionSafetyPolicyResponse> {
  try {
    return await httpClient.get(`${PREFIX}/session-policy`).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

export async function listPromptSafetyIntercepts(
  query?: ListSafetyInterceptsQuery,
): Promise<SafetyInterceptListResponse> {
  const params = new URLSearchParams()
  if (query?.category) params.set('category', query.category)
  if (query?.limit != null) params.set('limit', String(query.limit))
  if (query?.offset != null) params.set('offset', String(query.offset))
  const qs = params.toString()
  const url = qs ? `${PREFIX}/intercepts?${qs}` : `${PREFIX}/intercepts`
  try {
    return await httpClient.get(url).json()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}

File: server/chainup_agent/api/schemas/admin_prompt_safety.py
"""Pydantic DTOs — Admin prompt safety (ai.prompt-safety)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from chainup_agent.api.schemas.prompt_management import PromptPackSummaryOut

SafetyInterceptCategory = Literal["runtime", "prompt", "tool", "session"]
ToolEnforcementMode = Literal["default_deny", "allowlist_only", "approve_then_allow"]
ToolPolicyKind = Literal["deny", "allowlist", "shadow"]
RuntimeGovernanceHrefKind = Literal["confirmation", "policy", "routing"]


class SafetyBlocklistRuleOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    rule_id: str = Field(alias="ruleId")
    phrase: str
    match_mode: str = Field(alias="matchMode")


class SafetyBlocklistOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    safety_phrase_blocklist_revision: str = Field(alias="safetyPhraseBlocklistRevision")
    safety_phrase_scan_scope_default: str = Field(alias="safetyPhraseScanScopeDefault")
    rules: list[SafetyBlocklistRuleOut]


class SafetyOverviewOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    safety_phrase_blocklist_revision: str = Field(alias="safetyPhraseBlocklistRevision")
    safety_phrase_scan_scope_default: str = Field(alias="safetyPhraseScanScopeDefault")
    platform_safety_pack: PromptPackSummaryOut | None = Field(
        default=None, alias="platformSafetyPack"
    )
    safety_prompt_pack_count: int = Field(alias="safetyPromptPackCount")
    published_safety_pack_count: int = Field(alias="publishedSafetyPackCount")
    enabled_confirmation_rules_count: int = Field(alias="enabledConfirmationRulesCount")
    global_agent_switch_on: bool = Field(alias="globalAgentSwitchOn")
    ops_suspended: bool = Field(alias="opsSuspended")


class SafetyInterceptItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    ts: datetime
    category: SafetyInterceptCategory
    scenario_label: str = Field(alias="scenarioLabel")
    kind_label: str = Field(alias="kindLabel")
    reason: str
    subject: str | None = None
    execution_id: str | None = Field(default=None, alias="executionId")
    matched_rule_id: str | None = Field(default=None, alias="matchedRuleId")


class SafetyInterceptListResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    items: list[SafetyInterceptItem]
    total: int


class RuntimeSafetyGovernanceRow(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    key: str
    name: str
    risk_level_label: str = Field(alias="riskLevelLabel")
    auto_exec_summary: str = Field(alias="autoExecSummary")
    confirm_summary: str = Field(alias="confirmSummary")
    breaker_summary: str = Field(alias="breakerSummary")
    href_kind: RuntimeGovernanceHrefKind = Field(alias="hrefKind")


class RuntimeSafetyGovernanceResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    items: list[RuntimeSafetyGovernanceRow]


class ToolSafetyPolicyRow(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    key: str
    tool_pattern: str = Field(alias="toolPattern")
    policy: ToolPolicyKind
    enforcement_mode: ToolEnforcementMode = Field(alias="enforcementMode")
    note: str


class ToolSafetyPolicyResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    items: list[ToolSafetyPolicyRow]


class SessionSafetyPolicyItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    key: str
    label: str
    summary: str
    status: Literal["enabled", "warning", "info"]


class SessionSafetyPolicyResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    items: list[SessionSafetyPolicyItem]
